runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java

```diff
@@ -306,11 +308,19 @@ public void translateNode(
     @Override
     public void translateNode(
         Reshuffle<K, InputT> transform, FlinkBatchTranslationContext context) {
-
-      DataSet<WindowedValue<KV<K, InputT>>> inputDataSet =
+      final DataSet<WindowedValue<KV<K, InputT>>> inputDataSet =
           context.getInputDataSet(context.getInput(transform));
-
-      context.setOutputDataSet(context.getOutput(transform), inputDataSet.rebalance());
+      @SuppressWarnings("unchecked")
+      final CoderTypeInformation<WindowedValue<KV<K, InputT>>> outputType =
+          ((CoderTypeInformation) inputDataSet.getType())
+              .withPipelineOptions(context.getPipelineOptions());
+      final DataSet<WindowedValue<KV<K, InputT>>> retypedDataSet =
+          new MapOperator<>(
+              inputDataSet,
+              outputType,
+              FlinkIdentityFunction.of(),
```

Comment:
It looks like we could stick the existing patterns to load the options as part of the operator's open method.

Reply:
This wouldn't help, the MapOperator is called on "map side" of the rebalance. we encounter problems on "reduce side". Only reason for this `MapOperator` is, that we are unable to change TypeInformation of the input dataset.

However the main reason is, that DoFns, that follow rebalance are currently not chainable, therefore there is a slight chance of this happening.

I reckon that after resolving https://issues.apache.org/jira/browse/BEAM-8608, this would be no longer needed, because we'd init fs in chained DoFn's open method.

Reply:
I see. Would be good then to generally move to this approach. Let's add a comment here why this is necessary.